feelpp + clang-11 + boost1.71: qs_hdg_laplacian_2d failed to compile (debian bullseye)

For the records, latest feelpp dev FTBS on debian Bullseye with errors in quickstart laplacian hdg:
```
make[2]: *** [feelpp/quickstart/laplacian/CMakeFiles/feelpp_qs_hdg_laplacian_2d.dir/build.make:82: feelpp/quickstart/laplacian/CMakeFiles/feelpp_qs_hdg_laplacian_2d.dir/qs_hdg_laplacian.cpp.o] Error 1
make[2]: Target 'feelpp/quickstart/laplacian/CMakeFiles/feelpp_qs_hdg_laplacian_2d.dir/build' not remade because of errors.
make[1]: *** [CMakeFiles/Makefile2:42231: feelpp/quickstart/laplacian/CMakeFiles/feelpp_qs_hdg_laplacian_2d.dir/all] Error 2
Scanning dependencies of target feelpp_qs_laplacian_2dh
[ 95%] Building CXX object feelpp/quickstart/laplacian/CMakeFiles/feelpp_qs_laplacian_2dh.dir/qs_laplacian.cpp.o
In file included from /home/LNCMI-G/trophime/github/feelpp/feelpp/quickstart/laplacian/qs_laplacian.cpp:25:
In file included from /home/LNCMI-G/trophime/github/feelpp/feelpp/feel/../feel/feelcore/environment.hpp:46:
In file included from /home/LNCMI-G/trophime/github/feelpp/feelpp/contrib/pybind11/include/pybind11/pybind11.h:43:
In file included from /home/LNCMI-G/trophime/github/feelpp/feelpp/contrib/pybind11/include/pybind11/attr.h:13:
In file included from /home/LNCMI-G/trophime/github/feelpp/feelpp/contrib/pybind11/include/pybind11/cast.h:16:
/home/LNCMI-G/trophime/github/feelpp/feelpp/contrib/pybind11/include/pybind11/detail/internals.h:186:9: warning: 'PyEval_InitThreads' is deprecated [-Wdeprecated-declarations]
        PyEval_InitThreads();
        ^
/usr/include/python3.9/ceval.h:130:1: note: 'PyEval_InitThreads' has been explicitly marked deprecated here
Py_DEPRECATED(3.9) PyAPI_FUNC(void) PyEval_InitThreads(void);
^
/usr/include/python3.9/pyport.h:508:54: note: expanded from macro 'Py_DEPRECATED'
#define Py_DEPRECATED(VERSION_UNUSED) __attribute__((__deprecated__))
                                                     ^
/home/LNCMI-G/trophime/github/feelpp/feelpp/quickstart/laplacian/qs_hdg_laplacian.cpp:172:50: error: use of undeclared identifier 'f'
        auto u = cgLaplacian( cgXh, std::tuple{K,f,p_exact,un,r_1,r_2} );
                                                 ^
/home/LNCMI-G/trophime/github/feelpp/feelpp/quickstart/laplacian/qs_hdg_laplacian.cpp:172:60: error: use of undeclared identifier 'un'
        auto u = cgLaplacian( cgXh, std::tuple{K,f,p_exact,un,r_1,r_2} );
                                                           ^
/home/LNCMI-G/trophime/github/feelpp/feelpp/quickstart/laplacian/qs_hdg_laplacian.cpp:172:63: error: use of undeclared identifier 'r_1'; did you mean '_1'?
        auto u = cgLaplacian( cgXh, std::tuple{K,f,p_exact,un,r_1,r_2} );
                                                              ^~~
                                                              _1
/usr/include/boost/bind/placeholders.hpp:46:38: note: '_1' declared here
BOOST_STATIC_CONSTEXPR boost::arg<1> _1;
                                     ^
/home/LNCMI-G/trophime/github/feelpp/feelpp/quickstart/laplacian/qs_hdg_laplacian.cpp:172:67: error: use of undeclared identifier 'r_2'; did you mean '_2'?
        auto u = cgLaplacian( cgXh, std::tuple{K,f,p_exact,un,r_1,r_2} );
                                                                  ^~~
                                                                  _2
/usr/include/boost/bind/placeholders.hpp:47:38: note: '_2' declared here
BOOST_STATIC_CONSTEXPR boost::arg<2> _2;
                                     ^
/home/LNCMI-G/trophime/github/feelpp/feelpp/quickstart/laplacian/qs_hdg_laplacian.cpp:177:49: error: use of undeclared identifier 'locals'
                                        _inputs=locals
                                                ^
/home/LNCMI-G/trophime/github/feelpp/feelpp/quickstart/laplacian/qs_hdg_laplacian.cpp:208:33: error: use of undeclared identifier 'f'
                          _expr=f*id(w), _quad=ioption("rhs_quad") );
                                ^
/home/LNCMI-G/trophime/github/feelpp/feelpp/quickstart/laplacian/qs_hdg_laplacian.cpp:211:39: error: use of undeclared identifier 'un'
                          _expr=id(l)*un, _quad=ioption("rhs_quad")  );
                                      ^
/home/LNCMI-G/trophime/github/feelpp/feelpp/quickstart/laplacian/qs_hdg_laplacian.cpp:215:40: error: use of undeclared identifier 'r_2'; did you mean '_2'?
                           _expr=id(l)*r_2);
                                       ^~~
                                       _2
```

Note also the warning about python 3.9 pybind11. I guess this could be solved using the latest pybind11 release (2.6.2).

--- a/feelpp/quickstart/laplacian/qs_hdg_laplacian.cpp
+++ b/feelpp/quickstart/laplacian/qs_hdg_laplacian.cpp
@@ -173,7 +173,7 @@ int hdg_laplacian()
         Feel::cout << "-- CG<" << OrderP+1 << "> starts ----------------------------------------------------------\n";
         auto cgXh = Pch<OrderP+1>(mesh);
         Feel::cout << "cgXh<" << OrderP+1 << "> : " << cgXh->nDof() << std::endl;
-        auto u = cgLaplacian( cgXh, std::tuple{K,f,p_exact,un,r_1,r_2} );
+        auto u = cgLaplacian( cgXh, std::tuple{k,f,p_exact,un,r_1,r_2} );
         if ( u )        
             status_cg = check( checker( _name= "L2/H1 convergence cG", 
                                         _solution_key="p",

--- a/feelpp/quickstart/laplacian/qs_hdg_laplacian.cpp
+++ b/feelpp/quickstart/laplacian/qs_hdg_laplacian.cpp
@@ -174,12 +174,14 @@ int hdg_laplacian()
         auto cgXh = Pch<OrderP+1>(mesh);
         Feel::cout << "cgXh<" << OrderP+1 << "> : " << cgXh->nDof() << std::endl;
         auto u = cgLaplacian( cgXh, std::tuple{k,f,p_exact,un,r_1,r_2} );
+#if defined(FEELPP_HAS_SYMPY)
         if ( u )        
             status_cg = check( checker( _name= "L2/H1 convergence cG", 
                                         _solution_key="p",
                                         _gradient_key="grad_p",
                                         _inputs=locals
                                        ), *u );
+#endif
         Feel::cout << "-- CG<" << OrderP+1 << "> done ----------------------------------------------------------\n";
     }
     auto u = Vh->element( "u" );
@@ -381,23 +383,26 @@ int hdg_laplacian()
 
     tic();
 
+    int status1 = 0, status2 = 0, status3 = 0;
+#if defined(FEELPP_HAS_SYMPY)
     bool has_dirichlet = nelements(markedfaces(mesh,"Dirichlet"),true) >= 1;
     solution_t s_t = has_dirichlet?solution_t::unique:solution_t::up_to_a_constant;
-    int status1 = check( checker( _name= "L2/H1 convergence of potential", 
-                                  _solution_key="p",
-                                  _gradient_key="grad_p",
-                                  _inputs=locals
-                                ), pp, s_t );
-    int status2 = check( checker( _name= "L2 convergence of the flux", 
-                                  _solution_key="u",
-                                  _inputs=locals
-                                ), up );
-    int status3 = check( checker( _name= "L2/H1 convergence of postprocessed potential", 
-                                  _solution_key="p",
-                                  _gradient_key="grad_p",
-                                  _inputs=locals
-                                ), ppp, s_t );    
+    status1 = check( checker( _name= "L2/H1 convergence of potential", 
+                              _solution_key="p",
+                              _gradient_key="grad_p",
+                              _inputs=locals
+                              ), pp, s_t );
+    status2 = check( checker( _name= "L2 convergence of the flux", 
+                              _solution_key="u",
+                              _inputs=locals
+                              ), up );
+    status3 = check( checker( _name= "L2/H1 convergence of postprocessed potential", 
+                              _solution_key="p",
+                              _gradient_key="grad_p",
+                              _inputs=locals
+                              ), ppp, s_t );    
     // end::check[]
+#endif
 
     return status_cg || status1 || status2 || status3;
 }

--- a/feelpp/quickstart/laplacian/qs_hdg_laplacian.cpp
+++ b/feelpp/quickstart/laplacian/qs_hdg_laplacian.cpp
@@ -100,8 +100,6 @@ int hdg_laplacian()
     int proc_rank = Environment::worldComm().globalRank();
     auto Pi = M_PI;
     
-    auto K = expr(soption("k"));
-    auto lambda = cst(1.)/K;
     
 #if defined(FEELPP_HAS_SYMPY)
 
@@ -120,6 +118,7 @@ int hdg_laplacian()
     auto p_exact = expr( p_exact_str );
     auto u_exact = expr<FEELPP_DIM,1>( u_exact_str );
     auto k = expr( locals.at("k") );
+    auto lambda = cst(1.)/k;
     auto un = expr( locals.at("un") );
     auto f = expr( locals.at("f") );
     auto g = expr( locals.at("g") );
@@ -130,8 +129,13 @@ int hdg_laplacian()
     std::string u_exact_str = soption("solution.u");
     auto p_exact = expr(p_exact_str);
     auto u_exact = expr<Dim,1>(u_exact_str);
-    auto un_exact = trans(u_exact)*N();
-    auto f_exact = expr( soption( "functions.f") );
+    auto k = expr(soption("k"));
+    auto lambda = cst(1.)/k;
+    auto un = trans(u_exact)*N();
+    auto f = expr( soption( "functions.f") );
+    auto g = p_exact;
+    auto r_1 = cst(0.);
+    auto r_2 = un;
 #endif
     tic();
     auto mesh = loadMesh( new Mesh<Simplex<Dim>> );

--- a/feelpp/quickstart/laplacian/qs_hdg_laplacian_ibc.cpp
+++ b/feelpp/quickstart/laplacian/qs_hdg_laplacian_ibc.cpp
@@ -112,10 +112,7 @@ int hdg_laplacian()
 
     int proc_rank = Environment::worldComm().globalRank();
     auto Pi = M_PI;
-    
-    auto K = expr(soption("k"));
-    auto lambda = cst(1.)/K;
-    
+
 #if defined(FEELPP_HAS_SYMPY)
 
     std::map<std::string,std::string> inputs{{"dim",std::to_string(Dim)},{"k",soption("k")},{"p",soption("checker.solution")},{"grad_p",""}, {"u",""}, {"un",""}, {"f",""}, {"g",""}, {"J",""}, {"r_1",soption("r_1")}, {"r_2",soption("r_2")}};
@@ -133,12 +130,28 @@ int hdg_laplacian()
     auto p_exact = expr( p_exact_str );
     auto u_exact = expr<FEELPP_DIM,1>( u_exact_str );
     auto k = expr( locals.at("k") );
+    auto lambda = cst(1.)/k;
     auto un = expr( locals.at("un") );
     auto f = expr( locals.at("f") );
     auto g = expr( locals.at("g") );
     auto r_1 = expr( locals.at("r_1") );
     auto r_2 = expr( locals.at("r_2") ); 
     auto J_exact = expr( locals.at("J") );
+#else
+    std::string p_exact_str = soption("solution.p");
+    std::string u_exact_str = soption("solution.u");
+    auto p_exact = expr(p_exact_str);
+    auto u_exact = expr<Dim,1>(u_exact_str);
+    auto k = expr(soption("k"));
+    auto lambda = cst(1.)/k;
+    auto un = trans(u_exact)*N();
+    auto f = expr( soption( "functions.f") );
+    auto g = p_exact;
+    auto r_1 = cst(0.);
+    auto r_2 = un;
+    auto ibc_exact = expr( soption("functions.i") );
+    auto J_exact = inner(u_exact)/k;
+#endif
 
     std::map<std::string,double> ibc_exact_map;
     for( auto const& [ibc_type, ibc_data ] : ibcs )
@@ -146,16 +159,6 @@ int hdg_laplacian()
             ibc_exact_map[ibc_type] = 0;
         else
             ibc_exact_map[ibc_type] = integrate(markedfaces(mesh,ibc_data.second), un).evaluate()(0,0);
-
-#else
-    std::string p_exact_str = soption("solution.p");
-    std::string u_exact_str = soption("solution.u");
-    auto p_exact = expr(p_exact_str);
-    auto u_exact = expr<Dim,1>(u_exact_str);
-    auto un_exact = trans(u_exact)*N();
-    auto f_exact = expr( soption( "functions.f") );
-    auto ibc_exact = expr( soption("functions.i") );
-#endif
 
     // ****** Hybrid-mixed formulation ******
     // We treat Vh, Wh, and Mh separately
@@ -205,13 +208,15 @@ int hdg_laplacian()
     {
         auto cgXh = Pch<OrderP+1>(mesh);
         Feel::cout << "cgXh<" << OrderP+1 << "> : " << cgXh->nDof() << std::endl;
-        auto u = cgLaplacian( cgXh, std::tuple{K,f,p_exact,un,r_1,r_2} );
+        auto u = cgLaplacian( cgXh, std::tuple{k,f,p_exact,un,r_1,r_2} );
+#if defined(FEELPP_HAS_SYMPY)
         if ( u )        
             status_cg = check( checker( _name= "L2/H1 convergence cG", 
                                         _solution_key="p",
                                         _gradient_key="grad_p",
                                         _inputs=locals
                                        ), *u );
+#endif
     }
 
     auto u = Vh->element( "u" );
@@ -483,11 +488,11 @@ int hdg_laplacian()
     q.on( _range=elements(mesh), _expr=p_exact );
 
     
-    double I1 = integrate( _range=elements(mesh), _expr=K*gradv(pp)*trans(gradv(pp)), _quad=ioption("rhs_quad") ).evaluate()( 0,0 );
-    double I2 = integrate( _range=elements(mesh), _expr=inner(idv(up))/K, _quad=ioption("rhs_quad") ).evaluate()( 0,0 );
-    double I3 = integrate( _range=elements(mesh), _expr=inner(u_exact)/K, _quad=ioption("rhs_quad") ).evaluate()( 0,0 );
-    double I4 = integrate( _range=elements(mesh), _expr=K*inner(gradv(q)), _quad=ioption("rhs_quad") ).evaluate()( 0,0 );
-    double I5 = integrate( _range=elements(mesh), _expr=K*inner(gradv(ppp)), _quad=ioption("rhs_quad") ).evaluate()( 0,0 );
+    double I1 = integrate( _range=elements(mesh), _expr=k*gradv(pp)*trans(gradv(pp)), _quad=ioption("rhs_quad") ).evaluate()( 0,0 );
+    double I2 = integrate( _range=elements(mesh), _expr=inner(idv(up))/k, _quad=ioption("rhs_quad") ).evaluate()( 0,0 );
+    double I3 = integrate( _range=elements(mesh), _expr=inner(u_exact)/k, _quad=ioption("rhs_quad") ).evaluate()( 0,0 );
+    double I4 = integrate( _range=elements(mesh), _expr=k*inner(gradv(q)), _quad=ioption("rhs_quad") ).evaluate()( 0,0 );
+    double I5 = integrate( _range=elements(mesh), _expr=k*inner(gradv(ppp)), _quad=ioption("rhs_quad") ).evaluate()( 0,0 );
     double I6 = integrate( _range=elements(mesh), _expr=J_exact, _quad=ioption("rhs_quad") ).evaluate()( 0,0 );
 
     std::vector<double> J = {I1,I2,I5,I3,I4,I6};
@@ -522,26 +527,28 @@ int hdg_laplacian()
 
     tic();
 
+    int status1 = 0, status2 = 0, status3 = 0;
+#if defined(FEELPP_HAS_SYMPY)
     // tag::check[]
     bool has_dirichlet = nelements(markedfaces(mesh,"Dirichlet"),true) >= 1;
     solution_t s_t = has_dirichlet?solution_t::unique:solution_t::up_to_a_constant;
-    int status1 = check( checker( _name= "L2/H1 convergence of potential", 
-                                  _solution_key="p",
-                                  _gradient_key="grad_p",
-                                  _inputs=locals
-                                ), pp, s_t );
-    int status2 = check( checker( _name= "L2 convergence of the flux", 
-                                  _solution_key="u",
-                                  _inputs=locals
-                                ), up );
-    int status3 = check( checker( _name= "L2/H1 convergence of postprocessed potential", 
-                                  _solution_key="p",
-                                  _gradient_key="grad_p",
-                                  _inputs=locals
-                                ), ppp, s_t );    
+    status1 = check( checker( _name= "L2/H1 convergence of potential", 
+                              _solution_key="p",
+                              _gradient_key="grad_p",
+                              _inputs=locals
+                              ), pp, s_t );
+    status2 = check( checker( _name= "L2 convergence of the flux", 
+                              _solution_key="u",
+                              _inputs=locals
+                              ), up );
+    status3 = check( checker( _name= "L2/H1 convergence of postprocessed potential", 
+                              _solution_key="p",
+                              _gradient_key="grad_p",
+                              _inputs=locals
+                              ), ppp, s_t );    
     
     // end::check[]
-
+#endif
     return status_cg || status1 || status2 || status3;
 }
 

--- a/feelpp/quickstart/laplacian/qs_hdg_laplacian_ibc_dt.cpp
+++ b/feelpp/quickstart/laplacian/qs_hdg_laplacian_ibc_dt.cpp
@@ -116,10 +116,7 @@ int hdg_laplacian()
 
     int proc_rank = Environment::worldComm().globalRank();
     auto Pi = M_PI;
-    
-    auto K = expr(soption("k"));
-    auto lambda = cst(1.)/K;
-    
+
 #if defined(FEELPP_HAS_SYMPY)
 
     std::map<std::string,std::string> inputs{{"dim",std::to_string(Dim)},{"k",soption("k")},{"p",soption("checker.solution")},{"grad_p",""}, {"u",""}, {"un",""}, {"f",""}, {"g",""}, {"J",""}, {"r_1",soption("r_1")}, {"r_2",soption("r_2")}};
@@ -137,12 +134,27 @@ int hdg_laplacian()
     auto p_exact = expr( p_exact_str );
     auto u_exact = expr<FEELPP_DIM,1>( u_exact_str );
     auto k = expr( locals.at("k") );
+    auto lambda = cst(1.)/k;
     auto un = expr( locals.at("un") );
     auto f = expr( locals.at("f") );
     auto g = expr( locals.at("g") );
     auto r_1 = expr( locals.at("r_1") );
     auto r_2 = expr( locals.at("r_2") ); 
     auto J_exact = expr( locals.at("J") );
+#else
+    std::string p_exact_str = soption("solution.p");
+    std::string u_exact_str = soption("solution.u");
+    auto p_exact = expr(p_exact_str);
+    auto u_exact = expr<Dim,1>(u_exact_str);
+    auto k = expr(soption("k"));
+    auto lambda = cst(1.)/k;
+    auto un = trans(u_exact)*N();
+    auto f = expr( soption( "functions.f") );
+    auto g = p_exact;
+    auto r_1 = cst(0.);
+    auto r_2 = un;
+    auto J_exact = inner(u_exact)/k;
+#endif
 
     std::map<std::string,double> ibc_exact_map;
     for( auto const& [ibc_type, ibc_data ] : ibcs )
@@ -150,16 +162,6 @@ int hdg_laplacian()
             ibc_exact_map[ibc_type] = 0;
         else
             ibc_exact_map[ibc_type] = integrate(markedfaces(mesh,ibc_data.second), un).evaluate()(0,0);
-
-#else
-    std::string p_exact_str = soption("solution.p");
-    std::string u_exact_str = soption("solution.u");
-    auto p_exact = expr(p_exact_str);
-    auto u_exact = expr<Dim,1>(u_exact_str);
-    auto un_exact = trans(u_exact)*N();
-    auto f_exact = expr( soption( "functions.f") );
-    auto ibc_exact = expr( soption("functions.i") );
-#endif
 
     // ****** Hybrid-mixed formulation ******
     // We treat Vh, Wh, and Mh separately
@@ -224,12 +226,14 @@ int hdg_laplacian()
         auto cgXh = Pch<OrderP+1>(mesh);
         Feel::cout << "cgXh<" << OrderP+1 << "> : " << cgXh->nDof() << std::endl;
         auto u = cgLaplacian( cgXh, std::tuple{K,f,p_exact,un,r_1,r_2} );
+#if defined(FEELPP_HAS_SYMPY)
         if ( u )        
             status_cg = check( checker( _name= "L2/H1 convergence cG", 
                                         _solution_key="p",
                                         _gradient_key="grad_p",
                                         _inputs=locals
                                        ), *u );
+#endif
     }
 
     auto u = Vh->element( "u" );
@@ -567,6 +571,7 @@ int hdg_laplacian()
 
         tic();
 
+#if defined(FEELPP_HAS_SYMPY)
         // tag::check[]
         bool has_dirichlet = nelements(markedfaces(mesh,"Dirichlet"),true) >= 1;
         solution_t s_t = has_dirichlet?solution_t::unique:solution_t::up_to_a_constant;
@@ -585,9 +590,9 @@ int hdg_laplacian()
                                     _inputs=locals
                                     ), ppp, s_t );
         status_time |= (status1 || status2 || status3);
+        // end::check[]
+#endif
     }
-    
-    // end::check[]
 
     return status_cg || status_time;
 }

